storage/blockdevice/COMPONENT_SPINAND/include/SPINAND/SPINANDBlockDevice.h

```diff
@@ -0,0 +1,328 @@
+ *
+ *      printf("SPINAND BD size: %llu\n",         block_device.size());
+ *      printf("SPINAND BD read size: %llu\n",    block_device.get_read_size());
+ *      printf("SPINAND BD program size: %llu\n", block_device.get_program_size());
+ *      printf("SPINAND BD erase size (at address 0): %llu\n", sector_size_at_address_0);
+ *
+ *      // Write "Hello World!" to the first block
+ *      char *buffer = (char *) malloc(sector_size_at_address_0);
+ *      sprintf(buffer, "Hello World!\n");
+ *      block_device.erase(0, sector_size_at_address_0);
+ *      block_device.program(buffer, 0, sector_size_at_address_0);
+ *
+ *      // Read back what was stored
+ *      block_device.read(buffer, 0, sector_size_at_address_0);
+ *      printf("%s", buffer);
+ *
+ *      // Deinitialize the device
+ *      block_device.deinit();
+ *  }
+ *  @endcode
+ */
+class SPINANDBlockDevice : public mbed::BlockDevice {
+public:
+    /** Create SPINANDBlockDevice - An SPI NAND Flash Block Device over QSPI bus
+     *
+     *  @param io0 1st IO pin used for sending/receiving data during data phase of a transaction
+     *  @param io1 2nd IO pin used for sending/receiving data during data phase of a transaction
+     *  @param io2 3rd IO pin used for sending/receiving data during data phase of a transaction
+     *  @param io3 4th IO pin used for sending/receiving data during data phase of a transaction
+     *  @param sclk QSPI Clock pin
+     *  @param csel QSPI chip select pin
+     *  @param clock_mode specifies the QSPI Clock Polarity mode (SPINAND_POLARITY_MODE_0/SPINAND_POLARITY_MODE_1)
+     *         default value = 0
+     *  @param freq Clock frequency of the QSPI bus (defaults to 40MHz)
+     */
+    SPINANDBlockDevice(PinName io0 = MBED_CONF_SPINAND_SPINAND_IO0,
+                       PinName io1 = MBED_CONF_SPINAND_SPINAND_IO1,
+                       PinName io2 = MBED_CONF_SPINAND_SPINAND_IO2,
+                       PinName io3 = MBED_CONF_SPINAND_SPINAND_IO3,
+                       PinName sclk = MBED_CONF_SPINAND_SPINAND_SCK,
+                       PinName csel = MBED_CONF_SPINAND_SPINAND_CSN,
+                       int clock_mode = MBED_CONF_SPINAND_SPINAND_POLARITY_MODE,
+                       int freq = MBED_CONF_SPINAND_SPINAND_FREQ);
+
+    /** Initialize a block device
+     *
+     *  @return         SPINAND_BD_ERROR_OK(0) - success
+     *                  SPINAND_BD_ERROR_DEVICE_ERROR - device driver transaction failed
+     *                  SPINAND_BD_ERROR_READY_FAILED - Waiting for Memory ready failed or timedout
+     */
+    virtual int init();
+
+    /** Deinitialize a block device
+     *
+     *  @return         SPINAND_BD_ERROR_OK(0) - success
+     *                  SPINAND_BD_ERROR_DEVICE_ERROR - device driver transaction failed
+     */
+    virtual int deinit();
+
+    /** Desctruct SPINANDBlockDevie
+      */
+    ~SPINANDBlockDevice()
+    {
+        deinit();
```

Comment:
Do we need this object to be copyable or movable? We seem to support multiple flash devices using `csel` to select a different device. We also lock/unlock a mutex in the constructor and destructor. What happens if we start copying this object? Should we delete the copy constructor and copy-assignment operator? We have a helper class to delete the copy operations in `platform/include/platform/NonCopyable.h` if we think it's necessary. It's also possible we do need copy operations here and I'm missing something.